fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -946,6 +961,120 @@ public void rebuildIndex(@Nonnull FDBRecordStore store) {
+     * {@link #BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY}, {@link #BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY}, and
+     * {@link #BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY_OR_READABLE} are sorted in a way that each one is more exhaustive
+     * than the ones before it.
+     * <p>
+     * In which, {@link #BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY} is recommended if there is no reason to believe
+     * current index data is corrupted.
+     * </p>
+     */
+    public enum BuildOption {
+        /**
+         * Only build if the index is disabled.
+         */
+        BUILD_IF_DISABLED,
+        /**
+         * Build if the index is disabled; Continue build if the index is write-only.
+         * <p>
+         * Recommended. This should be sufficient if current index data is not corrupted.
+         * </p>
+         */
+        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY,
+        /**
+         * Build if the index is disabled; Rebuild if the index is write-only.
+         */
+        BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY,
+        /**
+         * Rebuild the index anyway.
+         */
+        BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY_OR_READABLE,
+        /**
+         * Build if the index is disabled; Continue build if the index is write-only, but rebuild if the index is
+         * readable.
+         * <p>
+         * This option should not be considered. It does not make much sense in any use case can be thought of. It is
+         * provided just for completeness.
+         * </p>
+         */
+        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_REBUILD_IF_READABLE,
+    }
+
+
+    /**
+     * Builds an index across multiple transactions safely. It differs from {@link #buildIndexAsync()} in two ways:
+     * <ul>
+     *     <li>
+     *         It stops with {@link com.apple.foundationdb.synchronizedsession.SynchronizedSessionLockedException} if
+     *         there is another runner actively working on the same index.
+     *     </li>
+     *     <li>
+     *         It checks and updates index states and clear index data respecting the {@link BuildOption}. So there is
+     *         no need to change the index states or clear existing index data before invoking this method.
+     *     </li>
+     * </ul>
+     * @return a future that will be ready when the build has completed.
+     * @throws com.apple.foundationdb.synchronizedsession.SynchronizedSessionLockedException the build is stopped
+     * because there may be another build running actively on this index.
+     */
+    // TODO: It should support range build (for building index in parallel) as well.
+    @Nonnull
+    public CompletableFuture<Void> safelyBuildIndexAsync() {
+        return runner
+                .runAsync(context -> openRecordStore(context).thenApply(store -> store.indexBuildLockSubspace(index)))
+                .thenCompose(lockSubspace -> runner.startSynchronizedSessionAsync(lockSubspace, DEFAULT_LEASE_LENGTH_MILLIS))
```

Comment:
I agree. I'm going to change this after deciding the one above.

Reply:
`SynchronizedSessionRunner` doesn't support changing the lease time on the fly yet so it might be misleading to put it `Config`. I've added it to `Builder` for now.